clojure/src/related/core.clj

```diff
@@ -0,0 +1,74 @@
+                                (let [post (nth posts i)
+                                      res  (reduce (fn [res tag]
+                                                     (update-in res [tag] conj i))
+                                                   res
+                                                   (:tags post))]
+                                  (recur (inc i) res))))
+
+          tagged-post-count (make-array Integer/TYPE n)
+
+          related           (->> posts
+                                 (map-indexed (fn [post-idx post]
+                                                (java.util.Arrays/fill tagged-post-count 0)
+
+                                                (let [top5 (make-array Integer/TYPE 10)]
+                                                  (doseq [tag (:tags post)
+                                                          idx (tag-map tag)]
+                                                    (aset-int tagged-post-count idx (inc (aget tagged-post-count idx))))
+
+                                                  (aset tagged-post-count post-idx 0)
+
+                                                  (loop [i        0
+                                                         min-tags 0]
+                                                    (if (< i n)
+                                                      (let [cnt (aget tagged-post-count i)]
+                                                        (if (> cnt min-tags)
+                                                          (let [up (loop [upper-bound 6]
+                                                                     (if-not (and (>= upper-bound 0)
+                                                                                  (> cnt (aget top5 upper-bound)))
+                                                                       upper-bound
+                                                                       (do
+                                                                         (aset top5 (+ upper-bound 2) (aget top5 upper-bound))
+                                                                         (aset top5 (+ upper-bound 3) (aget top5 (inc upper-bound)))
+                                                                         (recur (- upper-bound 2)))))]
+                                                            (aset top5 (+ up 2) cnt)
+                                                            (aset top5 (+ up 3) i)
+                                                            (recur (inc i) (aget top5 8)))
+                                                          (recur (inc i) min-tags)))))
+
+                                                  {:_id     (:_id post)
+                                                   :tags    (:tags post)
+                                                   :related [(nth posts (aget top5 1))
+                                                             (nth posts (aget top5 3))
```

Comment:
Manually indexing isn't allowed. Use integer division

```js
for (let i = 1; i < 10; i += 2) {
      topPosts[Math.floor(i / 2)] = posts[top5[i]];
}
```

or skip and append:

```crystal
(1...10).step(2) do |i|
    top_posts << posts[top5[i]]
end
```

Reply:
Why not? It's just a little optimization for not to create `range` or loop with its vars.

Reply:
So what exact rule this technique against? :)

Reply:
I didn't think of making a rule for this as it's a bad practice in general. I should be able to change 1 constant and get the topN without looking at the code. I know some of the solutions doesn't adhere to this but that could be done with a slight medication. 

ie: changing the 10 above to 2*topN

Reply:
I think it depends. It may be considering as an array initialization like e.g in Zig:

```
var top_5 = [5]u8{ 0, 0, 0, 0, 0 };
```

Anyway, it would be better to make some note in the rules, because making some optimizations and getting that it isn't allowed because it's a bad practice in general in the benchmark repo a little bit strange.

Reply:
I think it depends. It may be considering as an array initialization like e.g in Zig:
```
var top_5 = [5]u8{ 0, 0, 0, 0, 0 };
```
Anyway, it would be better to make some note in the rules, because making some optimizations and getting that it isn't allowed because it's a bad practice in general in the benchmark repo is a little bit strange.